content/gotraining.slide

```diff
@@ -740,11 +748,108 @@ Use `gofmt` or `goimport` as part of your editor/IDE setup
 [[https://godoc.org/gopkg.in/square/go-jose.v2][go-jose]] - JWT utils
  ...
 
-* Project work
+
+* A RESTful API starter (1/3)
+
+[[https://goplay.space/#oPHSAHa_Zo1][Full basic sample]]
+[[https://goplay.space/#ddCcD-1JdDJ][Full sample]] with error handling
+
+	type TeapotHandler interface {
+		handlePost(w http.ResponseWriter, r *http.Request)
+		handleGet(w http.ResponseWriter, r *http.Request)
+	}
+
+	func SetupRoutes(h TeapotHandler, r chi.Router) {
+		r.Post("/api/teapot", h.handlePost)
+		r.Get("/api/teapot/{id}", h.handleGet)
+	}
+
+	func main() {
+		r := chi.NewRouter()
+		th := NewMapTeapotHandler()
+		SetupRoutes(th, r)
+		http.ListenAndServe(":7735", r)
+	}
+
+
+* A RESTful API starter (2/3)
+
+	type Teapot struct {
+		ID          int
+		Colour      string
+		Temperature int
+	}
+
+	// MapTeapotHandler implements TeapotHandler
+	type MapTeapotHandler struct {
+		store map[int]Teapot
+		maxID int
+	}
+
+	func NewMapTeapotHandler() *MapTeapotHandler {
+		return &MapTeapotHandler{store: make(map[int]Teapot), maxID: 0}
```

Comment:
`return &MapTeapotHandler{store: map[int]Teapot{}}` - i have no real preference for make vs literal, but your boss does. And you can let `maxID` be initialised with the zero value. But I don't know if you intended for this to be explicit for a reason or not, so take it or leave it.

Reply:
Done.

Reply:
Ditto. Done in sample, not in slides.

Reply:
Donedone.